### PR TITLE
#1024 Hover state for palette dropdown aren't Carbon compliant

### PR DESCRIPTION
**Pull request by nabilIbm**
Fixes : #1024 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Review by tomlyn — CHANGES_REQUESTED** [review]
Please look at some of the items we discussed on webex.

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 1]
```diff
@@ -182,11 +183,12 @@ $palette-border-color: $ui-03;
...
+	max-height: 46px;
 	align-items: center;
 	justify-content: space-between;
 	display: flex;
-	border-bottom: $palette-border-width solid $palette-border-color;
+	margin-left: 0;
+	//border-bottom: $palette-border-width solid $palette-border-color;
 }
 
 .palette-flyout-category-caret {
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:194]: I think this ruleset can be removed.

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:194]: I think the palette-flyout-category-caret and palette-flyout-category-caret-closed  rulesets below, can be completely removed - right?

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:194]: I think the palette-flyout-category-caret and palette-flyout-category-caret-closed  rulesets below, can be completely removed - right?

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 2]
```diff
@@ -90,15 +90,15 @@ $palette-border-color: $ui-03;
...
 
 	.palette-list-item {
-		height: 46px;
+		height: 42px;
 		border-width: 0;
 		display: flex;
 		flex-direction: column;
 		align-items: center;
-		/* Prevent elements from being selectable */
-		/* This stops it being dragged incorrectly.*/
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:99]: The comment above is supposed to go with the `user-select: none;`  line of code. Can you put it back above that line please?

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:99]: The comment above is supposed to go with the `user-select: none;`  line of code. Can you put it back above that line please?

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 3]
```diff
@@ -182,11 +183,12 @@ $palette-border-color: $ui-03;
...
 	border-width: 0 0 0 $palette-border-width;
 	height: fit-content;
-	min-height: 46px;
+	max-height: 46px;
 	align-items: center;
 	justify-content: space-between;
 	display: flex;
-	border-bottom: $palette-border-width solid $palette-border-color;
+	margin-left: 0;
+	//border-bottom: $palette-border-width solid $palette-border-color;
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:191]: Please remove commented out code.

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:191]: Please remove commented out code.

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 4]
```diff
@@ -206,9 +208,10 @@ $palette-border-color: $ui-03;
 .palette-flyout-category-item {
 	display: flex;
 	align-items: center;
+	user-select: none;
 	/* Prevent elements from being selectable */
 	/* This stops it being dragged incorrectly.*/
-	user-select: none;
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:211]: Why was this removed. It prevents text from being selected (highlighted) by accident. Unless there is a good reason can you put it back please. Also, the comment above goes with the line of code so it the line of code is removed the comment also has to be removed.

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:211]: Why was this removed. It prevents text from being selected (highlighted) by accident. Unless there is a good reason can you put it back please. Also, the comment above goes with the line of code so it the line of code is removed the comment also has to be removed.

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 5]
```diff
@@ -217,8 +220,10 @@ $palette-border-color: $ui-03;
 
 .palette-flyout-category-item-icon {
 	width: 20px;  /* Not specifying height preserves the image's aspect ratio. */
-	margin-left: 16px;
+	margin-left: 5px;
 	fill: $icon-01;
+	padding-top: 0rem;
+	margin-right: 0.2rem;
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:226]: We generally use pixel (px) sizes in the SCSS rather than rem units. Is there some reason you are using rem here? Can you use px here if possible please?

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:226]: We generally use pixel (px) sizes in the SCSS rather than rem units. Is there some reason you are using rem here? Can you use px here if possible please?

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 6]
```diff
@@ -233,30 +238,26 @@ $palette-border-color: $ui-03;
...
 	line-height: 46px;  /* Must be same as containing div height */
 	margin-left: 16px;
-	width: 35px;
 }
 
 .palette-flyout-category-text-no-image {
 	line-height: 46px;  /* Must be same as containing div height */
 	margin-left: 16px;
 	width: 170px;
+
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:253]: Please remove unwanted blank line.

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:253]: Please remove unwanted blank line.

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 7]
```diff
@@ -265,6 +266,7 @@ $palette-border-color: $ui-03;
 
 		// Move the magnifying glass icon to be positioned over the category icons
 		// in the palette, as required from DUX review.
+
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:269]: Remove blank line

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:269]: Remove blank line

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 8]
```diff
@@ -306,6 +308,12 @@ $palette-border-color: $ui-03;
 	position: absolute;
 	overflow-x: hidden;
 	overflow-y: overlay;
+	li .bx--accordion__arrow {
+		margin-top: 0.2rem;
+	}
+	button::before .common-canvas.classic-vertical .palette-flyout-categories .tooltip-container{
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:314]: This selector is not valid. The use of .common-canvas.classic-vertical  will only work in the test harness. In general we should not have to use such long selectors to do the styling. What is this even doing?

Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:314]: This selector is not valid. The use of .common-canvas.classic-vertical  will only work in the test harness. In general we should not have to use such long selectors to do the styling. What is this even doing?

**Review by tomlyn — CHANGES_REQUESTED** [review]
Hi Nabil, I have reviewed some, but not all of your changes in this PR. Also, as you know, I sent you in Slack a set of (9) things that need to be fixed. I think its probably best if you get those fixed before I proceed with any more reviewing since you will undoubtedly have to make more changes. Thanks.

File: canvas_modules/common-canvas/src/palette/palette.scss (hunk 1 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette.scss:194]: Correct !! Accordion takes care of them.

**Review by tomlyn — CHANGES_REQUESTED** [review]
I will wait until you've worked on the 'loading categories' code before reviewing further.

File: canvas_modules/common-canvas/__tests__/palette/palette-flyout-test.js [hunk 9]
```diff
@@ -176,10 +176,14 @@ describe("Palette renders correctly", () => {
...
+		// We now click the Import category again to close it
 		const importCat2 = findCategoryElement(wrapper, "Import");
 		importCat2.simulate("click");
-		expect(wrapper.find(PaletteFlyoutContentList)).to.have.length(1);
-		expect(wrapper.find(PaletteContentListItem)).to.have.length(2);
+		// When the Import category is closed and the Outputs category is opened
+		// we should have 2 palettes flyout content lists objects because
+		// content lists always exist whether categories are opened or closed.
+		expect(wrapper.find(PaletteFlyoutContentList)).to.have.length(2);
+		expect(wrapper.find(PaletteContentListItem)).to.have.length(5);
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/__tests__/palette/palette-flyout-test.js:186]: Acually i did the other way around. i test the only category is opened which is "Outputs" and "Import" already closed
image

File: canvas_modules/harness/cypress/support/canvas/verification-cmds.js [hunk 10]
```diff
@@ -868,6 +868,12 @@ Cypress.Commands.add("verifyNodeDoesExistInPaletteAtIndex", (nodeName, index) =>
 		.then((idx) => expect(idx).to.equal(index));
 });
 
+Cypress.Commands.add("verifyCategoryIsClosed", (categoryName) => {
+	cy.get("li.bx--accordion__item.bx--accordion__item--active .bx--accordion__title .palette-flyout-category")
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/harness/cypress/support/canvas/verification-cmds.js:872]: Fixed !!

File: canvas_modules/harness/cypress/support/canvas/node-cmds.js [hunk 11]
```diff
@@ -519,10 +519,15 @@ Cypress.Commands.add("hoverOverCategory", (nodeCategory) => {
 	cy.findCategory(nodeCategory).trigger("mouseover");
 });
 
-Cypress.Commands.add("findNodeInCategory", (nodeLabel) => {
+Cypress.Commands.add("findNodeInCategory", (nodeLabel, paletteName) => {
 	cy.document().then((doc) => {
-		// Palette Layout - Modal
-		if (doc.canvasController.getCanvasConfig().enablePaletteLayout === "Modal") {
+		if (paletteName) {
+			cy.get("li.bx--accordion__item.bx--accordion__item--active #accordion-item-28")
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/harness/cypress/support/canvas/node-cmds.js:525]: Fixed !! updating the unique id# with  class ".bx--accordion__content"

File: canvas_modules/common-canvas/__tests__/palette/palette-flyout-test.js [hunk 12]
```diff
@@ -162,8 +162,8 @@ describe("Palette renders correctly", () => {
 		const wrapper = createMountedPalette();
 		const importCat = findCategoryElement(wrapper, "Import");
 		importCat.simulate("click");
-		expect(wrapper.find(PaletteFlyoutContentList)).to.have.length(1);
-		expect(wrapper.find(PaletteContentListItem)).to.have.length(3);
+		expect(wrapper.find(PaletteFlyoutContentList)).to.have.length(2);
+		expect(wrapper.find(PaletteContentListItem)).to.have.length(5);
```
Comment (tomlyn) [inline canvas_modules/common-canvas/__tests__/palette/palette-flyout-test.js:166]: The code in this test which looks for PaletteFlyoutContentList and PaletteContentListItem are not testing if the categories are open or closed.  Can we use the code from the  test below this one 
```
	const categoryList = wrapper.find("div.palette-flyout-categories");
	expect(categoryList.find(".bx--accordion__item--active")).to.have.length(2);
```
to check how many categories are open? 

Perhaps you could add these two lines to a separate function and then call it from all the places where we want to check how many categories are open.

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 13]
```diff
@@ -95,9 +93,9 @@ $palette-border-color: $ui-03;
 		display: flex;
 		flex-direction: column;
 		align-items: center;
-		/* Prevent elements from being selectable */
-		/* This stops it being dragged incorrectly.*/
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:99]: Requesting for a third time that you put back these comments and position them above the `user-select: none;`  line.

**Review by tomlyn — CHANGES_REQUESTED** [review]
I made some more comments. Please also fix the `show` prop as we discussed on Slack.

File: canvas_modules/common-canvas/src/palette/palette.scss (hunk 13 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette.scss:99]: I did put the comment back

File: canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx [hunk 14]
```diff
@@ -127,7 +107,6 @@ class PaletteFlyoutContent extends React.Component {
 				canvasController={this.props.canvasController}
 				isPaletteOpen={this.props.isPaletteOpen}
 				isEditingEnabled={this.props.isEditingEnabled}
-				// isShowRanking // Uncomment this to show ranking for debuggig ranking algorithm
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx:130]: Fixed

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 15]
```diff
@@ -524,3 +546,7 @@ $palette-border-color: $ui-03;
 .palette-content-list {
 	width: 100%;
 }
+
+.palette-loading-category .bx--accordion__arrow {
+	display: none;
+}
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette.scss:552]: Fixed.

File: canvas_modules/common-canvas/src/palette/palette-content-list.jsx [hunk 16]
```diff
@@ -58,15 +59,11 @@ class PaletteContentList extends React.Component {
...
-
 		const style = {};
-		style.borderBottom = this.props.isLastCategory ? "none" : null;
 		style.display = this.props.show ? "block" : "none";
 
 		return (
-			<div width="100%" draggable="false" className="palette-content-list palette-scroll"
-				style={ style }
-			>
+			<div width="100%" style={style} draggable="false" className="palette-content-list palette-scroll">
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette-content-list.jsx:66]: This is now changed as result of the show prop change.

File: canvas_modules/common-canvas/src/palette/palette-content-list.jsx [hunk 17]
```diff
@@ -76,11 +73,11 @@ class PaletteContentList extends React.Component {
...
 	category: PropTypes.object.isRequired,
 	nodeTypeInfos: PropTypes.array.isRequired,
-	show: PropTypes.bool.isRequired,
 	canvasController: PropTypes.object.isRequired,
 	isPaletteOpen: PropTypes.bool.isRequired,
-	isLastCategory: PropTypes.bool.isRequired,
-	isEditingEnabled: PropTypes.bool.isRequired
+	isEditingEnabled: PropTypes.bool.isRequired,
+	show: PropTypes.bool.isRequired,
+
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette-content-list.jsx:80]: Both show & isLastCategory have to be removed.

File: canvas_modules/common-canvas/__tests__/palette/palette-flyout-test.js (hunk 12 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/__tests__/palette/palette-flyout-test.js:166]: categoriesOpenCounts function has been created !!

File: canvas_modules/harness/cypress/support/canvas/node-cmds.js [hunk 18]
```diff
@@ -519,10 +519,15 @@ Cypress.Commands.add("hoverOverCategory", (nodeCategory) => {
 	cy.findCategory(nodeCategory).trigger("mouseover");
 });
 
-Cypress.Commands.add("findNodeInCategory", (nodeLabel) => {
+Cypress.Commands.add("findNodeInCategory", (nodeLabel, categoryLabel) => {
 	cy.document().then((doc) => {
-		// Palette Layout - Modal
-		if (doc.canvasController.getCanvasConfig().enablePaletteLayout === "Modal") {
+		if (categoryLabel) {
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/harness/cypress/support/canvas/node-cmds.js:524]: Done !!

File: canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx [hunk 19]
```diff
@@ -127,7 +108,6 @@ class PaletteFlyoutContent extends React.Component {
...
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx:130]: In the earlier comment I asked you to put this line back in and your response says "Fixed" -- but it is not there?! What happened? I don't understand how you can say it is Fixed when it isn't.

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 20]
```diff
@@ -168,57 +168,89 @@ $palette-border-color: $ui-03;
...
+		button {
+			border-bottom: $palette-border-width solid $palette-border-color;
+		}
+	}
+	.bx--accordion__item {
+		border-top: 0;
+
+		.bx--accordion__heading {
+			padding: 0;
+			// border-bottom: $palette-border-width solid $palette-border-color;
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:202]: Why is this here?

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 21]
```diff
@@ -168,57 +168,89 @@ $palette-border-color: $ui-03;
...
 			.palette-list-item-icon-and-text {
 				min-height: 30px;
+				padding-left: 10px;
 			}
 		}
 	}
+}
 
+.palette-flyout-categories {
+	height: calc(100% - 48px);
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette.scss:184]: I realized this 48px should really be 41px which is the height assigned to palette-flyout-search-container. Please change this to be:
```
height: calc(100% - $palette-search-container-height);
```
(And see my associated comment below about making 41px into a variable for this file.)

**Review by tomlyn — CHANGES_REQUESTED** [review]
I added a few more comments. In addition to fixing the things I commented on can you also do the following. Copy the file `harness/src/client/components/custom-canvases/flows/flowsLoadingPalette.json`    into the `harness/test_resources/palettes` directory and rename it to be `loadingPalette.json`.  Going forward that will give us a way of easily testing the loading categories.

File: canvas_modules/common-canvas/src/palette/palette.scss [hunk 22]
```diff
@@ -253,16 +287,11 @@ $palette-border-color: $ui-03;
 	padding-left: 5px;
 }
 
-.palette-flyout-category:hover {
-	font-weight: 600;
-}
-
 .palette-flyout-search-container {
-	height: 100%;
+	height: 41px;
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette.scss:291]: set the height sass variable

File: canvas_modules/common-canvas/src/palette/palette.scss (hunk 20 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette.scss:202]: Removed the comment !!

File: canvas_modules/common-canvas/__tests__/palette/palette-flyout-test.js [hunk 23]
```diff
@@ -291,3 +306,8 @@ function findCategoryElement(flyoutPaletteContent, categoryName) {
 	}
 	return null;
 }
+
+function categoriesOpenCounts(wrapper) {
+	const categoryList2 = wrapper.find("div.palette-flyout-categories");
+	return categoryList2.find(".bx--accordion__item--active");
+}
```
[the remark this answers is not in the document]
Comment (nabilIbm) [reply canvas_modules/common-canvas/__tests__/palette/palette-flyout-test.js:313]: Updated the function's name !!

File: canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx (hunk 19 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx:130]: Did put it back !!

File: canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx [hunk 24]
```diff
@@ -71,41 +72,21 @@ class PaletteFlyoutContent extends React.Component {
...
 					<PaletteFlyoutContentCategory
 						key={category.id}
 						category={category}
-						isCategorySelected={isCategorySelected}
 						categorySelectedMethod={this.categorySelected}
-						itemCount={nodeTypeInfos.length}
 						canvasController={this.props.canvasController}
 						isPaletteOpen={this.props.isPaletteOpen}
+						isEditingEnabled={this.props.isEditingEnabled}
+						// isShowRanking // Uncomment this to show ranking for debuggig ranking algorithm.
```
Comment (tomlyn) [inline canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx:84]: This is the wrong place

**Review by tomlyn — CHANGES_REQUESTED** [review]

**Review by tomlyn — COMMENTED** [review]
Move isRanking commented code -- as we discussed

File: canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx (hunk 24 above)
Comment (nabilIbm) [reply canvas_modules/common-canvas/src/palette/palette-flyout-content.jsx:84]: I removed it from here and pasted it in the previous place, where it was in the first place !!

**Review by tomlyn — CHANGES_REQUESTED** [review]

**Review by tomlyn — CHANGES_REQUESTED** [review]
Re-requesting changes

**Review by tomlyn — DISMISSED** [review]

**Review by tomlyn — APPROVED** [review]
